bubuku/features/rebalance/change.py

```diff
@@ -0,0 +1,248 @@
+                if weight_list:
+                    candidates[(source_broker_id, target_broker_id)] = sorted(
+                        weight_list, key=lambda x: x[1], reverse=True)
+        return candidates
+
+    @staticmethod
+    def _place_to_cardinalities(array: list, topic: str, cardinality_change: int):
+        old_cardinality = None
+        for item in [(a, b) for a, b in array if a == topic]:
+            old_cardinality = item[1]
+            array.remove(item)
+        if old_cardinality is None:
+            return
+        new_cardinality = old_cardinality + cardinality_change
+        idx = len(array)
+        for i in range(idx - 1, -1, -1):
+            if array[i][1] >= new_cardinality:
+                idx = i + 1
+                break
+        array.insert(idx, (topic, new_cardinality))
+
+    @staticmethod
+    def rebalance_leaders(brokers: dict) -> list:
+        candidates = OptimizedRebalanceChange._build_leader_distribution_weights(brokers)
+        queue_list = []
+        while any([broker.have_extra_leaders() for broker in brokers.values()]):
+            top_candidate = None
+            for broker_pair, weight_list in candidates.items():
+                if not top_candidate:
+                    top_candidate = broker_pair
+                else:
+                    if weight_list[0][1] > candidates[top_candidate][0][1]:
+                        top_candidate = broker_pair
+            # Taking best topic to move
+            topic, _ = candidates[top_candidate][0]
+            # Updating cardinality map
+            for broker_pair, weight_list in candidates.items():
+                if top_candidate == broker_pair:
+                    OptimizedRebalanceChange._place_to_cardinalities(weight_list, topic, -2)
+                elif top_candidate[0] == broker_pair[0] or top_candidate[1] == broker_pair[1]:  # src or dst matches
+                    OptimizedRebalanceChange._place_to_cardinalities(weight_list, topic, -1)
+            # Get broker objects
+            source_broker, target_broker = brokers[top_candidate[0]], brokers[top_candidate[1]]
+
+            # Selecting best partition to move (It is better to just swap leadership, instead of copying data)
+            selected_partition = None
+            source_partitions = source_broker.list_partitions(topic, False)
+            target_partitions = target_broker.list_partitions(topic, True)
+            for partition in source_partitions:
+                if partition in target_partitions:
+                    selected_partition = partition
+                    break
+            # Failed to find optimized partition for swapping, taking any.
+            selected_partition = source_partitions[0] if selected_partition is None else selected_partition
+            topic_partition = (topic, selected_partition)
+            # Update brokers state
+            target_broker.accept_leader(source_broker, topic_partition)
```

Comment:
Is it correct to say that the guarantee that this method works is in this line? Since for every iteration we always call `accept_leader` which has side effects by removing something from `source_broker` and adding to `target_broker`.